### PR TITLE
Implements transformation-based synthesis in Q#

### PR DESCRIPTION
**Pull request by msoeken**
This is an Q# implementation of the unidirectional variant of the transformation based synthesis approach [Miller, Maslov, Dueck, DAC 2003].

There are two example operations that are using the automatic synthesis algorithms. The first example finds the circuit for the permutation and checks whether it correctly computes the permutation using exhaustive simulation of all input patterns. The second example implements an instance of the hidden-shift quantum algorithm, in which the bent function is the inner product applied to a two set of qubits, where the first set of qubits is permuted.

There are tests for the utility functions and the synthesis function.

I tried to follow code conventions and documentation styles as I saw it in other samples.

### CODE REVIEW

File: LibraryTests/Samples/ReversibleLogicSynthesisTests.qs [hunk 1]
```diff
@@ -0,0 +1,82 @@
+// Author: Mathias Soeken, EPFL (Mail: mathias.soeken@epfl.ch)
```
Comment (cgranade) [inline LibraryTests/Samples/ReversibleLogicSynthesisTests.qs]: I apologize for the hassle, but I believe that we need for the copyright header to match the headers on other files; would you be willing to modify the PR accordingly? Thanks!

File: Samples/ReversibleLogicSynthesis/ReversibleLogicSynthesis.qs [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+    // The first value is an array of qubits for the control lines, the second value
+    // is a qubit for the target line.
+    //
+    // The target cannot be contained in the control lines.
+    newtype MCTGate = (Qubit[], Qubit);
+
+    // # Summary
+    // Constructs a MCMTMask type as a singleton array if targets is not 0,
+    // otherwise returns an empty array
+    function MakeGateMask(controls: Int, targets: Int) : MCMTMask[] {
```
[the remark this answers is not in the document]
Comment (msoeken) [reply Samples/ReversibleLogicSynthesis/ReversibleLogicSynthesis.qs]: I will rename it to `GateMask`.

File: Samples/ReversibleLogicSynthesis/ReversibleLogicSynthesis.qs [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+        return Map(UpdateOutputPattern(_, gateMask), perm);
+    }
+
+    // # Summary
+    // Computes gate masks to transform perm[x] to x and updates the current permutation
+    function TBSStep(state : (Int[], MCMTMask[]), x : Int) : (Int[], MCMTMask[]) {
+        let (perm, gates) = state;
+        let y = perm[x];
+        let masks = GateMasksForAssignment(x, y);
+        let new_perm = Fold(UpdatePermutation, perm, masks);
```
[the remark this answers is not in the document]
Comment (msoeken) [reply Samples/ReversibleLogicSynthesis/ReversibleLogicSynthesis.qs]: Thanks!

File: Samples/ReversibleLogicSynthesis/ReversibleLogicSynthesis.qs [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+        return GateMasksToToffoliGates(qubits, masks);
+    }
+
+    ////////////////////////////////////////////////////////////
+    // Generic permutation synthesis                          //
+    ////////////////////////////////////////////////////////////
+
+    // # Summary
+    // Synthesize Toffoli network from a permutation using functional synthesis
+    operation PermutationOracle(perm : Int[], qubits : Qubit[], synth : ((Int[], Qubit[]) -> MCTGate[])) : () {
```
[the remark this answers is not in the document]
Comment (msoeken) [reply Samples/ReversibleLogicSynthesis/ReversibleLogicSynthesis.qs]: Makes sense, also makes the code in `HiddenShiftProblem` look more consistent.

File: LibraryTests/Samples/ReversibleLogicSynthesisTests.qs (hunk 1 above)
Comment (msoeken) [reply LibraryTests/Samples/ReversibleLogicSynthesisTests.qs]: Okay, will fix that.

**Comment by msoeken** [comment]
My last commit addresses the comments.

**Pull request merged**